yjit/src/codegen.rs

```diff
@@ -1657,10 +1645,51 @@ fn gen_expandarray(
 
     let comptime_recv = jit.peek_at_stack(&asm.ctx, 0);
 
-    // If the comptime receiver is not an array, bail
-    if comptime_recv.class_of() != unsafe { rb_cArray } {
-        gen_counter_incr(asm, Counter::expandarray_comptime_not_array);
-        return None;
+    // If the comptime receiver is not an array
+    if !unsafe { RB_TYPE_P(comptime_recv, RUBY_T_ARRAY) } {
+        // at compile time, ensure to_ary is not defined
+        let target_cme = unsafe { rb_callable_method_entry_or_negative(comptime_recv.class_of(), ID!(to_ary)) };
+        let cme_def_type = unsafe { get_cme_def_type(target_cme) };
+
+        if cme_def_type != VM_METHOD_TYPE_UNDEF {
+            // if to_ary is defined, return can't compile so to_ary can be called
+            return None;
```

Comment:
Could you leave a counter for this exit? Maybe rename the counter from `expandarray_comptime_not_array`.